lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+    llvm::SmallVector<Type> innerTypes;
+    structType.getInnerTypes(innerTypes);
+    return b.create<hw::StructExplodeOp>(loc, innerTypes, value).getResults();
+  }
+
+  llvm::SmallVector<Value> toBits(Value v, std::optional<StringRef> name = {}) {
+    llvm::SmallVector<Value> bits;
+    for (unsigned i = 0, e = v.getType().getIntOrFloatBitWidth(); i != e; ++i)
+      bits.push_back(b.create<comb::ExtractOp>(loc, v, i, /*bitWidth=*/1));
+    return bits;
+  }
+
+  ///  OR-reduction of the bits in 'v'.
+  Value rOr(Value v, std::optional<StringRef> name = {}) {
+    return buildNamedOp([&]() { return bOr(toBits(v)); }, name);
+  }
+
+  ///  Extract bits v[hi:lo] (inclusive).
+  Value extract(Value v, unsigned lo, unsigned hi,
+                std::optional<StringRef> name = {}) {
+    unsigned width = hi - lo + 1;
+    return buildNamedOp(
+        [&]() { return b.create<comb::ExtractOp>(loc, v, lo, width); }, name);
+  }
+
+  ///  Truncates 'value' to its lower 'width' bits.
+  Value truncate(Value value, unsigned width,
+                 std::optional<StringRef> name = {}) {
+    return extract(value, 0, width - 1, name);
+  }
+
+  Value zext(Value value, unsigned outWidth,
+             std::optional<StringRef> name = {}) {
+    unsigned inWidth = value.getType().getIntOrFloatBitWidth();
+    assert(inWidth <= outWidth && "zext: input width must be <- output width.");
+    if (inWidth == outWidth)
+      return value;
+    auto c0 = constant(outWidth - inWidth, 0);
+    return concat({c0, value}, name);
+  }
+
+  Value sext(Value value, unsigned outWidth,
+             std::optional<StringRef> name = {}) {
+    return comb::createOrFoldSExt(loc, value, b.getIntegerType(outWidth), b);
+  }
+
+  ///  Extracts a single bit v[bit].
+  Value bit(Value v, unsigned index, std::optional<StringRef> name = {}) {
+    return extract(v, index, index, name);
+  }
+
+  ///  Creates a hw.array of the given values.
+  Value arrayCreate(ValueRange values, std::optional<StringRef> name = {}) {
+    return buildNamedOp(
+        [&]() { return b.create<hw::ArrayCreateOp>(loc, values); }, name);
+  }
+
+  ///  Extract the 'index'th value from the input array.
+  Value arrayGet(Value array, Value index, std::optional<StringRef> name = {}) {
+    return buildNamedOp(
+        [&]() { return b.create<hw::ArrayGetOp>(loc, array, index); }, name);
+  }
+
+  ///  Muxes a range of values.
+  ///  The select signal is expected to be a decimal value which selects
+  ///  starting from the lowest index of value.
+  Value mux(Value index, ValueRange values,
+            std::optional<StringRef> name = {}) {
+    if (values.size() == 2)
+      return b.create<comb::MuxOp>(loc, index, values[1], values[0]);
+
+    return arrayGet(arrayCreate(values), index, name);
+  }
+
+  ///  Muxes a range of values. The select signal is expected to be a 1-hot
+  ///  encoded value.
+  Value ohMux(Value index, ValueRange inputs) {
```

Comment:
`ohMux` is overly short. `oneHotMux`? It's not that much worse with decent autocomplete.